### PR TITLE
Make message twig template discoverable

### PR DESCRIPTION
**Pull request by mccrodp**
Cleaner version of original PR: https://github.com/Gizra/message/pull/43

### CODE REVIEW

File: message.module [hunk 1]
```diff
@@ -285,3 +287,71 @@ function message_config_translation_info_alter(&$info) {
...
+  $sanitized_view_mode = strtr($variables['elements']['#view_mode'], '.', '_');
+
+  $suggestions[] = 'message__' . $sanitized_view_mode;
+  $suggestions[] = 'message__' . $message->bundle();
+  $suggestions[] = 'message__' . $message->bundle() . '__' . $sanitized_view_mode;
+  $suggestions[] = 'message__' . $message->id();
+  $suggestions[] = 'message__' . $message->id() . '__' . $sanitized_view_mode;
+
+  return $suggestions;
+}
```
Comment (RoySegall) [inline message.module]: Need a new line in the end of the file.

**Comment by RoySegall** [comment]
When rendering the message like this:

``` php
    $foo = Message::load(1);

    $bar = \Drupal::entityTypeManager()->getViewBuilder('message')->view($foo);
```

and put in the twig file the text foo I don't see it in the renderable array.

**Comment by mccrodp** [comment]
@RoySegall I added a new test `MessageTemplateSuggestionsTest` which uses `entityTypeManager` call similar to the below. Now that I have added a fix here for #35 I can verify I now see the render array. This may solve some other open issues in the process.

All tests should be green, so this should be ready / almost ready now, please check! :)

> When rendering the message like this:
> 
>    $foo = Message::load(1);
> 
>    $bar = \Drupal::entityTypeManager()->getViewBuilder('message')->view($foo);
> and put in the twig file the text foo I don't see it in the renderable array.

EDIT: If all is well, I think this means that https://github.com/Gizra/message/issues/32 can be closed now too.

**Comment by mccrodp** [comment]
@RoySegall - Template is now working, the content variable is populated and field are displayed as expected. I verified through work on Message UI :)

**Comment by RoySegall** [comment]
OK. Can you add another test for verify that when we do 

```
$foo = Message::load(1);
$bar = \Drupal::entityTypeManager()->getViewBuilder('message')->view($foo);
// Use actually the service but you got what i mean :)
render($bar);
```

we get the text wrapped with the div?

**Comment by mccrodp** [comment]
Ok, I'll add a new `testMessageTextWrapper()` to `MessageUiTest` that does something like:

```
public function testMessageTextWrapper() {
   $type = 'dummy_message';
    // Create message to be rendered.
    $message_type = $this->createMessageType($type, 'Dummy message', '', array('[message:author:name]'));
    $message = Message::create(array('type' => $message_type->id()))
      ->setAuthorId($this->user->id());

    $message->save();

    $this->drupalGet($message->urlInfo()); /*May not be necessary*/

    // Simulate theming of the message.
    $view_builder = \Drupal::entityTypeManager()->getViewBuilder($message>getEntityTypeId());
    $output = \Drupal::service('renderer')->render($view_builder->view($message));

    $result = $this->xpath(); /*check if xpath returns a div on $output*/
    $this->assertEqual($result, $output); /*return TRUE / FALSE if xpath returns a div on $output*/
  }
```

**Comment by RoySegall** [comment]
You could also use simple `strpos`

**Comment by mccrodp** [comment]
Need to fix this test, not sure where this pass by reference error is happening when I step through it.

Might try to use `strpos` instead of the xpath as I presume it has something to do with the call to setting raw render content or the xpath method.

**Comment by mccrodp** [comment]
I kept the `xpath` method and fixed this by correcting the use of view builder. @RoySegall - I think this is ready...?

File: src/Tests/MessageUiTest.php [hunk 2]
```diff
@@ -119,6 +120,27 @@ public function testMessageTranslate() {
...
+
+    $message->save();
+
+    // Simulate theming of the message.
+    $build = \Drupal::entityTypeManager()->getViewBuilder('message')->view($message);
+    $output = \Drupal::service('renderer')->renderRoot($build);
+    $this->setRawContent($output);
+
+    $xpath = $this->xpath('//div');
+    $this->assertTrue($xpath, 'A div has been found wrapping the message text.');
```
Comment (RoySegall) [inline src/Tests/MessageUiTest.php:141]: You can just do `strpos("<div>MESSAGE_TEXT", $output) !== FALSE`. Would be much more simple.

**Comment by mccrodp** [comment]
Yes, I agree and this is also a better test as it forces a check on `MESSAGE_TEXT`. However, there is a deeper issue, that has been there for a while, with retrieving the message text during render that I will have to fix in the process\* :cry: I guess it needs to be fixed at some stage though.
- At the moment it is checking on `<div>\n \n</div>` as there is this message text issue.

**Comment by RoySegall** [comment]
You don't need to get the message text from the render. you have the `getText()` method.

**Comment by mccrodp** [comment]
Ahhh, yes, very true! Great, I'll try get this updated today and then look at the message text issue during render at a later time.

**Comment by mccrodp** [comment]
So, I believe this will still have the same issue:

`strpos("<div>" . $message->getText(), $output) !== FALSE`

`$output` is from render not including the message text and `$message->getText()` will return the correct / expected message text and this operation will return `FALSE`.

I could just do `strpos("<div>", $output) !== FALSE` for the moment if that was acceptable?

**Comment by RoySegall** [comment]
Hmm... probably. Let me see the most optional solution about this one. Either way i'll emerge this one tonight.

**Comment by mccrodp** [comment]
Ok, I'll just get this green again and then leave it with you.

I wonder if it's related to `template_preprocess_message` and if the partials need to be explicitly loaded into `#markup` in there also. It's the only difference I can see from the render workflow compared to node.

**Comment by mccrodp** [comment]
@RoySegall - I reverted back to using xpath method as `string` property on the `$output` object is protected and the `__toString()` method throws an error message.

I think this is good to go as-is, unless you can improve this evening.

On a related note, in terms of the locating where the message text in `#markup` gets wiped, I've located it down to `\Drupal\Core\Render\Renderer::doRender()` line: 517 in this `Markup::create` method. I haven't gone through exactly why yet, but it exists before this call and afterwards it is overwritten with `\n \n`

`$elements['#markup'] = Markup::create($prefix . $elements['#children'] . $suffix);`

**Comment by RoySegall** [comment]
Nope. Still not work as i expected. When trying to run 

``` php
    $message = Message::load(1);

    $build = \Drupal::entityTypeManager()->getViewBuilder('message')->view($message);
    $output = \Drupal::service('renderer')->renderPlain($build);
```

adsadasd___site-install

And when trying to run 

``` php
\Drupal::theme()->render('message', ['elements' => ['#message' => $message]]);
```

I get 

> The website encountered an unexpected error. Please try again later.
> LogicException: Render context is empty, because render() was called outside of a renderRoot() or renderPlain() call. Use renderPlain()/renderRoot() or #lazy_builder/#pre_render instead. in Drupal\Core\Render\Renderer->doRender() (line 246 of core/lib/Drupal/Core/Render/Renderer.php).
> Drupal\Core\Render\Renderer->render(Array)
> Drupal\Core\Template\TwigExtension->escapeFilter(Object, Array, 'html', NULL, 1)
> __TwigTemplate_f0349ef89a49cf9dfa18a7a3a7d8cc090d647396c1d3f34cdf47385abcd122a2->doDisplay(Array, Array)
> Twig_Template->displayWithErrorHandling(Array, Array)
> Twig_Template->display(Array)
> Twig_Template->render(Array)
> twig_render_template('modules/message/templates/message.html.twig', Array)
> Drupal\Core\Theme\ThemeManager->render('message', Array)
> Drupal\devel\EventSubscriber\DevelEventSubscriber->onRequest(Object, 'kernel.request', Object)
> Drupal\Component\EventDispatcher\ContainerAwareEventDispatcher->dispatch('kernel.request', Object)
> Symfony\Component\HttpKernel\HttpKernel->handleRaw(Object, 1)
> Symfony\Component\HttpKernel\HttpKernel->handle(Object, 1, 1)
> Drupal\Core\StackMiddleware\Session->handle(Object, 1, 1)
> Drupal\Core\StackMiddleware\KernelPreHandle->handle(Object, 1, 1)
> Drupal\page_cache\StackMiddleware\PageCache->pass(Object, 1, 1)
> Drupal\page_cache\StackMiddleware\PageCache->handle(Object, 1, 1)
> Drupal\Core\StackMiddleware\ReverseProxyMiddleware->handle(Object, 1, 1)
> Drupal\Core\StackMiddleware\NegotiationMiddleware->handle(Object, 1, 1)
> Stack\StackedHttpKernel->handle(Object, 1, 1)
> Drupal\Core\DrupalKernel->handle(Object)

We need a way to render the message using API function(maybe theme or even apply a new method on the class) so this would help us in message ui and message notify and impact on message subscribe as well.

**Comment by mccrodp** [comment]
@RoySegall Can you give me an example of what is expected on the first issue? 

Also, where are you running both of these code snippets from?

```
LogicException: Render context is empty, because render() was called outside of a renderRoot() or renderPlain() call.
```

This 2nd issue is as it says I believe, you are running a render out of context and if so you need to run renderRoot as I do in MessageUiTest in this PR.
E.g.

```
$build = \Drupal::entityTypeManager()->getViewBuilder('message')->view($message);
$output = \Drupal::service('renderer')->renderRoot($build);
```

**Comment by RoySegall** [comment]
I'm running it from `DevelEventSubscriber::onRequest` which is the devel_init in D8. What I would expect is to run the code and get the render message content.

**Comment by mccrodp** [comment]
Ok, not sure how to use that, but I'm sure I can figure it out. I think you probably have a valid issue with the 1st problem, so I'll take a look. 

The 2nd error is a valid error I believe and would also happen with any other entity when calling that render method out of context. So it seems anyways.

**Comment by mccrodp** [comment]
I added the below to `DevelEventSubscriber::onRequest` directly.

```
$message = Message::load(2);

$build = \Drupal::entityTypeManager()->getViewBuilder('message')->view($message);
$output = \Drupal::service('renderer')->renderPlain($build);

dpm($output);
```

I could see your issue, there are 2 reasons:
1. As mentioned before, there is an issue with the display of the partials as explained previously and in https://github.com/Gizra/message/pull/48#issuecomment-173540603 We said we would fix this later
2. Fields need to be set to display, changing them from their hidden default at: `/admin/structure/message/manage/example_create_node/display`

I did then get the following template from `dpm()` as expected: https://gist.github.com/mccrodp/322835a0253b2ed01d33

**Comment by RoySegall** [comment]
Enable the `kint` module.

**Comment by mccrodp** [comment]
Ahh, thanks! Was wondering how to get pretty output. 

Won't be looking at this again for today. But I hope it can get merged and then look at any remaining issues after. As far as I'm aware they are not directly related to the pr code itself.

**Comment by RoySegall** [comment]
This is the last issue before the stable release.

**Comment by amitaibu** [comment]
Exciting :) but please don't tag a final release yet...

**Comment by RoySegall** [comment]
Of course.

**Comment by amitaibu** [comment]
:+:

**Comment by amitaibu** [comment]
:+1:

**Comment by mccrodp** [comment]
That sounds great. Is there anything left in this PR before it gets merged then?

Is the partials issue (https://github.com/Gizra/message/pull/48#issuecomment-173540603) fix is required before we tag a release? If so, should we open a separate PR, or what is best?

**Comment by RoySegall** [comment]
> That sounds great. Is there anything left in this PR before it gets merged then?

Yes. This PR need to fix the issue with the message render. I don't see any change since last Friday, when I checked it. In this PR i need to be able to use the theme function to render the message in different view modes and get the text of the message.

**Comment by mccrodp** [comment]
When you say "text of the message", you mean the partial? The fields are working fine for me, they just need to be enabled for display in the Manage Display.

I believe this is an issue separate to any code I have introduced in this PR, it's to do with the partials and render arrays in general, not any template related code. The `#markup` index of the build array is being overwritten as highlighted (by code / line no.) in https://github.com/Gizra/message/pull/48#issuecomment-173540603

This PR was originally just about making "...the template discoverable" and I believe it has done that. I won't get to look at this partials issue for about 2 weeks, so I recommend to merge this and open a new issue / PR for the partials render problem above. That means someone else can easily contribute a fix for this separately, your call! Thanks

**Pull request merged**